railgun/test/performance/perf_knative_test.go

```diff
@@ -0,0 +1,258 @@
+//+build performance_tests
+
+package performance
+
+import (
+	"crypto/tls"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"time"
+
+	"bytes"
+	"context"
+	"fmt"
+	"testing"
+
+	types "k8s.io/apimachinery/pkg/types"
+
+	"github.com/kong/kubernetes-testing-framework/pkg/kind"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	knativenetworkingversioned "knative.dev/networking/pkg/client/clientset/versioned"
+	"knative.dev/pkg/apis"
+	knservingv1 "knative.dev/serving/pkg/apis/serving/v1"
+	knativeversioned "knative.dev/serving/pkg/client/clientset/versioned"
+)
+
+const (
+	knativeCrds = "https://github.com/knative/serving/releases/download/v0.13.0/serving-crds.yaml"
+	knativeCore = "https://github.com/knative/serving/releases/download/v0.13.0/serving-core.yaml"
+)
+
+func TestPerfKnativePerformance(t *testing.T) {
+
+	proxy := KongInfo.ProxyURL.Hostname()
+	assert.NotEmpty(t, proxy)
+	t.Logf("proxy url %s", proxy)
+
+	ctx := context.Background()
+
+	t.Log("Deploying all resources that are required to run knative")
+	require.NoError(t, deployManifest(knativeCrds, ctx, t))
+	require.NoError(t, deployManifest(knativeCore, ctx, t))
+	require.True(t, isKnativeReady(ctx, cluster, t), true)
+
+	t.Log("Configure Knative NetworkLayer as Kong")
+	require.NoError(t, perfconfigKnativeNetwork(ctx, cluster, t))
+	require.NoError(t, perfconfigKnativeDomain(ctx, proxy, cluster, t))
+
+	cnt := 1
+	cost := 0
+	for cnt <= max_ingress {
+		namespace := fmt.Sprintf("knative-%d", cnt)
+		err := CreateNamespace(ctx, namespace, t)
+		assert.NoError(t, err)
+
+		t.Logf("Install knative service into namespace %s", namespace)
+		require.Eventually(t, func() bool {
+			err := perfInstallKnativeSrv(ctx, t, namespace)
+			if err != nil {
+				t.Logf("checking knativing webhook readiness.")
+				return false
+			}
+			return true
+		}, 30*time.Second, 2*time.Second, true)
+
+		t.Logf("Test knative service using kong within namespace %s.", namespace)
+		require.True(t, perfaccessKnativeSrv(ctx, proxy, t, namespace, &cost), true)
+		cnt++
+	}
+	t.Logf("knative ingress cost %d", cost/max_ingress)
+
+	// cleanup clean ingress also to continueally repeat
+}
+
+func deployManifest(yml string, ctx context.Context, t *testing.T) error {
+	cmd := exec.CommandContext(ctx, "kubectl", "apply", "-f", yml)
+	stdout, stderr := new(bytes.Buffer), new(bytes.Buffer)
+	cmd.Stdout = stdout
+	cmd.Stderr = stderr
+	if err := cmd.Run(); err != nil {
+		fmt.Fprintln(os.Stdout, stdout.String())
+		return err
+	}
+	t.Logf("successfully deploy manifest " + yml)
+	return nil
+}
```

Comment:
this is duplicate functionality from the integration tests, we should move it to some utility package and deduplicate, preferably under `internal/` so it doesn't become part of our public API.

Reply:
agree, add as a TODO.